File: python/complex-numbers/complex_numbers.py
import math

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        real = self.real * other.real - self.imaginary * other.imaginary
        imaginary = self.imaginary * other.real + self.real * other.imaginary
        return ComplexNumber(real, imaginary)

    def __sub__(self, other):
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __truediv__(self, other):
        real = (self.real * other.real + self.imaginary * other.imaginary) \
            / (other.real * other.real + other.imaginary * other.imaginary)
        imaginary = (self.imaginary * other.real - self.real * other.imaginary) \
            / (other.real * other.real + other.imaginary * other.imaginary)
        return ComplexNumber(real, imaginary)

    def __abs__(self):
        return math.sqrt(self.real * self.real + self.imaginary * self.imaginary)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        return ComplexNumber(math.exp(self.real) * math.cos(self.imaginary),
            math.exp(self.real) * math.sin(self.imaginary))

File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double real;
    private readonly double imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        this.real = real;
        this.imaginary = imaginary;
    }

    public double Real() => real;

    public double Imaginary() => imaginary;

    public ComplexNumber Mul(ComplexNumber other)
    {
        return new ComplexNumber(
            real * other.Real() - imaginary * other.Imaginary(),
            imaginary * other.Real() + real * other.Imaginary()
        );
    }

    public ComplexNumber Add(ComplexNumber other)
    {
        return new ComplexNumber(real + other.Real(), imaginary + other.Imaginary());
    }

    public ComplexNumber Sub(ComplexNumber other)
    {
        return new ComplexNumber(real - other.Real(), imaginary - other.Imaginary());
    }

    public ComplexNumber Div(ComplexNumber other)
    {
        double denominator = Math.Pow(other.Real(), 2) + Math.Pow(other.Imaginary(), 2);
        return new ComplexNumber(
            (real * other.Real() + imaginary * other.Imaginary()) / denominator,
            (imaginary * other.Real() - real * other.Imaginary()) / denominator
        );
    }

    public double Abs()
    {
        return Math.Sqrt(real * real + imaginary * imaginary);
    }

    public ComplexNumber Conjugate()
    {
        return new ComplexNumber(real, -imaginary);
    }
    
    public ComplexNumber Exp()
    {
        double eToTheReal = Math.Pow(Math.E, real);
        return new ComplexNumber(eToTheReal * Math.Cos(imaginary), eToTheReal * Math.Sin(imaginary));
    }
}